File: combinations.py
def get_combinations(strand, length):
  def combinations(base, array, current_depth):
    if current_depth < length:
      result = []
      for index, unit in enumerate(array):
        remaining = array[:index] + array[index+1:]
        result.extend(combinations(base + unit, remaining, current_depth + 1))
      return result
    return [base]

  strand_list = list(strand)
  result = []
  for index, unit in enumerate(strand_list):
    remaining = strand_list[:index] + strand_list[index+1:]
    result.extend(combinations(unit, remaining, 1))
  
  return result

File: combinations.js
const getCombinations = (strand, length) => {
  const depth = 1;
  
  const combinations = (base, array, currentDepth) => array
    .flatMap((unit, index, arr) => currentDepth < length
      ? combinations(base + unit, arr.toSpliced(index, 1), currentDepth + 1)
      : base + unit)

  return strand
    .split('')
    .flatMap((unit, index, array) => combinations(unit, array.toSpliced(index, 1), depth + 1));
}

const result = getCombinations('hack', 3)
console.log(result)
